persistence/jdbc/src/test/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedClusterTest.java

```diff
@@ -0,0 +1,84 @@
+import org.infinispan.persistence.jdbc.UnitTestDatabaseManager;
+import org.infinispan.persistence.jdbc.configuration.JdbcStringBasedStoreConfigurationBuilder;
+import org.infinispan.persistence.manager.PersistenceManager;
+import org.infinispan.test.TestingUtil;
+import org.infinispan.util.concurrent.CompletionStages;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "persistence.jdbc.JdbcStringBasedClusterTest")
+public class JdbcStringBasedClusterTest extends BaseDistStoreTest<Integer, String, JdbcStringBasedClusterTest> {
+
+   private ControlledTimeService controlledTimeService = new ControlledTimeService();
+
+   {
+      INIT_CLUSTER_SIZE = 2;
+      l1CacheEnabled = false;
+   }
+
+   @Override
+   protected void amendCacheManagerBeforeStart(EmbeddedCacheManager ecm) {
+      TestingUtil.replaceComponent(ecm, TimeService.class, controlledTimeService, true);
+   }
+
+   @Override
+   public Object[] factory() {
+      return new Object[]{
+            new JdbcStringBasedClusterTest().segmented(true).shared(true).cacheMode(CacheMode.DIST_SYNC),
+            new JdbcStringBasedClusterTest().segmented(false).shared(true).cacheMode(CacheMode.DIST_SYNC),
+            new JdbcStringBasedClusterTest().segmented(true).cacheMode(CacheMode.LOCAL),
+            new JdbcStringBasedClusterTest().segmented(false).cacheMode(CacheMode.LOCAL),
+      };
+   }
+
+   @Override
+   protected StoreConfigurationBuilder addStore(PersistenceConfigurationBuilder persistenceConfigurationBuilder, boolean shared) {
+      JdbcStringBasedStoreConfigurationBuilder storeBuilder = persistenceConfigurationBuilder
+            .addStore(JdbcStringBasedStoreConfigurationBuilder.class);
+      storeBuilder.shared(shared);
+      storeBuilder.segmented(segmented);
+      UnitTestDatabaseManager.configureUniqueConnectionFactory(storeBuilder);
+      UnitTestDatabaseManager.buildTableManipulation(storeBuilder.table());
+
+      return storeBuilder;
+   }
+
+   public void testPurgeExpired() {
+      Cache<String, String> cacheToUse = cache(0, cacheName);
+      cacheToUse.put("key1", "expired", 10, TimeUnit.MINUTES);
+      cacheToUse.put("key2", "value");
+      cacheToUse.put("key3", "expired", 10, TimeUnit.MINUTES);
+      cacheToUse.put("key4", "value");
+
+      assertEquals(4, cache(0, cacheName).size());
+
+      // Remove all entries from memory
+      for (int i = 0; i < cacheManagers.size(); ++i) {
+         TestingUtil.extractComponent(cache(i, cacheName), InternalDataContainer.class).clear();
+      }
+
+      assertEquals(4, cacheToUse.size());
+
+      // Advance to expire one of the entries
+      controlledTimeService.advance(TimeUnit.MINUTES.toMillis(11));
+
+      PersistenceManager persistenceManager = TestingUtil.extractComponent(cache(0, cacheName), PersistenceManager.class);
+      CompletionStages.join(persistenceManager.purgeExpired());
+
+      assertEquals(2, cacheToUse.size());
```

Comment:
Because of my concerns with keeping the `getBinaryStream()`, I would like it if the test was extended to 
* Put multiple keys in the same segment
* Verify the expiration listeners being invoked
* Verify the size of the store including expired entries (cache.size() skips expired entries)

Reply:
> Because of my concerns with keeping the `getBinaryStream()`, I would like it if the test was extended to
> 
> * Put multiple keys in the same segment

In the same segment makes no difference for the JDBC store, in fact purging doesn't even look at the segments.

> * Verify the expiration listeners being invoked

We already do that in `BaseNonBlockingStoreTest#testPurgeExpired`. My guess is the behavior of the getBinaryStream is implementation dependent and the ResultSet documentation was just trying to be overly protective, but obviously our test only tests H2, so I agree it was needed.

> * Verify the size of the store including expired entries (cache.size() skips expired entries)

You are thinking from the perspective to verify that the purge actually removed the entries? I can add that that to the new test by doing an explicit select.

Reply:
> We already do that in BaseNonBlockingStoreTest#testPurgeExpired.

Ah but the value wasn't verified, let me update that test.